### PR TITLE
[develop] Add GDAS and GEFS datasets to the workflow

### LINKED ISSUE #485: Start ensemble members from global ensemble members
## Description
The SRW App supports running a forecast ensemble whose perturbations are stochastically driven through several physical means. There needs to be a mechanism for perturbing initial and boundary conditions, as well. 

RRFS development has achieved this by using GEFS and GDAS ensemble members for initial and lateral boundary conditions. Those external models have been included for use with the retrieve_data.py tool. The workflow needs to be modified to collect the requested global ensemble members, prepare them as initial or boundary conditions using the make_ics and make_lbcs tasks, and have the forecast use those.

## Solution
- Modify the configuration layer to accept GEFS and GDAS as ICS/LBCS.
- Modify get_external_* task to get the requested members
- Modify the XML to run make_ics and make_lbcs on each global ensemble member.
- Modify the forecast tasks to ensure it uses the appropriate ICS/LBCS from the respective ensemble member.

## Requirements**
- Don't break existing capabilities to run with ICS/LBCS of a deterministic external model.
- Map the external global ensemble to the same SRW ensemble member. For example, RRFS member 1 corresponds to GEFS member 1, etc. 
- Grib files from GEFS come with both a & b files. They must be concatenated when they are retrieved. This should not be done as part of retreive_data.py as it's outside the scope of the tool, and should be included in the ex-script for the task.

## Acceptance Criteria (Definition of Done)
- ICS can be any of GDAS (netcdf atm+sfc files), GEFS (grib a & b), or any existing options
- LBCS can be any of GDAS (netcdf atm files), GEFS (merged grib a & b), or any existing options.
- GEFS a & b grib files are merged, if used.

### PR DESCRIPTION
**Pull request by EdwardSnyder-NOAA**
## DESCRIPTION OF CHANGES: 

This PR adds the ensemble (GDAS and GEFS) datasets as options for the SRW App and is part of a bigger effort to merge in RRFS capabilities into the develop branch.

Currently, the workflow ensemble tasks all use the same initial conditions. Because of the addition of the ensemble datasets, each ensemble task can now have its own initial conditions with the global ensemble number mapping to the SRW App ensemble number (i.e. GDAS mem002 corresponds to SRW App mem002). 

The appropriate logic was added to these files so that the GEFS and GDAS datasets can run with the workflow without breaking existing functionality. A number of changes were made in the `retrieve_data.py` script to handle the `--members` argument and GEFS filename list. A GEFS function was added to the `exregional_get_extrn_mdl_files.sh` file to merge the GEFS files and update the summary `extrn_mdl_var_defns.sh` script.

Testing was done on Jet and ran through experiments that used GEFS, GDAS, and FV3GFS as the initial conditions.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## TESTS CONDUCTED: 

- [ ] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [X] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DOCUMENTATION:

Updated the ConfigWorkflow and InputOutputFiles rst files to reflect the addition of GEFS and GDAS as ICS/LBCS datasets.

## ISSUE: 

This resolves issue: #485 - Start ensemble members from global ensemble members

## CHECKLIST

- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [X] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 

A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [X] enhancement
- [ ] documentation
- [ ] release
- [ ] high priority
- [ ] run_ci
- [ ] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [ ] help wanted

### CODE REVIEW

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 1]
```diff
@@ -56,14 +56,14 @@ or lateral boundary conditions for the FV3.
 #
 set -x
 if [ "${ICS_OR_LBCS}" = "ICS" ]; then
-  if [ ${TIME_OFFSET_HRS} -eq 0 ] ; then
+  if [ ${TIME_OFFSET_HRS} -eq 0 ] || [ "${EXTRN_MDL_NAME}" == "GDAS" ] ; then
```
Comment (christinaholtNOAA) [inline scripts/exregional_get_extrn_mdl_files.sh]: I don't follow this logic completely. Why would we never want to allow getting a forecast file from GDAS?

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 2]
```diff
@@ -171,6 +177,49 @@ ${cmd}
...
 #
+# Merge GEFS files
+#
+#-----------------------------------------------------------------------
+#
+if [ "${EXTRN_MDL_NAME}" = "GEFS" ]; then
+    for num in $(seq -f "%02g" ${NUM_ENS_MEMBERS}); do
+        if [ "${ICS_OR_LBCS}" = "LBCS" ]; then
+            fcst_hrs_tmp=( $fcst_hrs )
+            fcst_hrs_ary=( $(seq ${fcst_hrs_tmp[0]} ${fcst_hrs_tmp[2]} ${fcst_hrs_tmp[1]}) )
```
Comment (christinaholtNOAA) [inline scripts/exregional_get_extrn_mdl_files.sh]: Could this be spelled out as ...my brain took SO LONG with "what's an 'ary'?" I definitely paused on "second half of 'temporary'?" for too long! Haha.

I might also suggest `all_fcst_hrs_array` just to be clearer that we're expanding the (min, step, max) structure of the original array.

For efficiency, `fcst_hrs_tmp` and `fcst_hrs_ary` can be computed once outside the loop.

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 3]
```diff
@@ -171,6 +177,49 @@ ${cmd}
...
+                fn_list_1=( "gep$num.t${hh}z.pgrb2a.0p50.f`printf %03d $fcst_hr`" \
+                        "gep$num.t${hh}z.pgrb2b.0p50.f`printf %03d $fcst_hr`" \
+                        "gep$num.t${hh}z.pgrb2.0p50.f`printf %03d $fcst_hr`" )
+                fn_list_2=( "gep$num.t${hh}z.pgrb2af`printf %02d $fcst_hr`" \
+                        "gep$num.t${hh}z.pgrb2bf`printf %02d $fcst_hr`" \
+                        "gep$num.t${hh}z.pgrb2`printf %02d $fcst_hr`" )
+                fn_list_3=( "gep$num.t${hh}z.pgrb2af`printf %03d $fcst_hr`" \
+                        "gep$num.t${hh}z.pgrb2bf`printf %03d $fcst_hr`" \
+                        "gep$num.t${hh}z.pgrb2`printf %03d $fcst_hr`" )
+                fn_lists=( "fn_list_1" "fn_list_2" "fn_list_3" )
```
Comment (christinaholtNOAA) [inline scripts/exregional_get_extrn_mdl_files.sh]: Does it make sense to read this info in from the summary file so we don't duplicate information between here and the data_locations.yaml?

File: scripts/exregional_make_ics.sh [hunk 4]
```diff
@@ -389,6 +391,35 @@ case "${EXTRN_MDL_NAME_ICS}" in
...
+  input_type="grib2"
+  convert_nst=False
+  vgtyp_from_climo=True
+  sotyp_from_climo=True
+  vgfrc_from_climo=True
+  minmax_vgfrc_from_climo=True
+  lai_from_climo=True
+  tg3_from_soil=False
+  ;;
+
```
Comment (christinaholtNOAA) [inline scripts/exregional_make_ics.sh:422]: I am assuming these settings come straight from RRFS_dev1? I am seeing that the GDAS external_model=GFS in the latest version [here](https://github.com/NOAA-GSL/regional_workflow/blob/feature/RRFS_dev1/scripts/exregional_make_ics.sh).

Maybe it's a version discrepancy in chgres_cube, though?

All others look consistent, though.

File: scripts/exregional_make_lbcs.sh [hunk 5]
```diff
@@ -291,6 +293,20 @@ case "${EXTRN_MDL_NAME_LBCS}" in
   fi
   ;;
 
+"GDAS")
+  tracers_input="[\"spfh\",\"clwmr\",\"o3mr\",\"icmr\",\"rwmr\",\"snmr\",\"grle\"]"
+  tracers="[\"sphum\",\"liq_wat\",\"o3mr\",\"ice_wat\",\"rainwat\",\"snowwat\",\"graupel\"]"
+  external_model="GFS"
+  input_type="gaussian_netcdf"
+  fn_atm="${EXTRN_MDL_FNS[0]}"
+  ;;
```
Comment (christinaholtNOAA) [inline scripts/exregional_make_lbcs.sh:302]: It looks like we may need additional logic to allow LBCS to be nemsio format given the namelist entries in [RRFS_dev1](https://github.com/NOAA-GSL/regional_workflow/blob/feature/RRFS_dev1/scripts/exregional_make_lbcs.sh).

That could be a part of a future PR, though.

File: ush/retrieve_data.py [hunk 6]
```diff
@@ -363,9 +364,12 @@ def get_requested_files(cla, file_templates, input_locs, method="disk", **kwargs
...
     for mem in members:
-        target_path = fill_template(cla.output_path, cla.cycle_date, mem=mem)
-        target_path = create_target_path(target_path)
-
+        if isinstance(mem, int):
+            output_path = f"{cla.output_path}/mem{mem:03d}"
+        else:
+            output_path = cla.output_path
+        target_path = fill_template(output_path, cla.cycle_date, mem=mem)
+        target_path = create_target_path(target_path, cla, mem)
```
Comment (christinaholtNOAA) [inline ush/retrieve_data.py]: I think it might be best to require the user to provide the full path template to the output directory of choice so that we can keep this tool as generic as possible. For example, if the user set's 

```
output_path = '/some/place/on/disk/mem{mem:03d}'
```
... then none of this logic needs to be ensemble specific, or modify a path at all. It simplifies the tool and makes the App make its own choices.

**Comment by MichaelLueken** [comment]
@EdwardSnyder-NOAA Do you still have the modified config files to run an experiment with either GDAS or GEFS data available on Jet?  While attempting to retrieve GDAS NetCDF files, I see the following in the log file:

```
/apps/hpss/hsi
INFO: Checking hpss for GDAS

+ '[' GDAS = GEFS ']'
End exregional_get_extrn_mdl_files.sh at Wed Jan 11 15:36:38 UTC 2023 with error code 0 (time elapsed: 00:00:00)
```
It ends without pulling data from HPSS, causing the `make_ics` and `make_lbcs` tasks to fail.  While attempting to pull GEFS grib2 data from AWS, I see the following in the log file:

```
INFO: Checking aws for GEFS

INFO: Getting files named like [['gep{mem:02d}.t{hh}z.pgrb2a.0p50.f{fcst_hr:03d}', 'gep{mem:02d}.t{hh}z.pgrb2b.0p50.f{fcst_hr:03d}'], ['gep{mem:02d}.t{hh}z.pgrb2af{fcst_hr:02d}', 'gep{mem:02d}.t{hh}z.pgrb2bf{fcst_hr:02d}'], ['gep{mem:02d}.t{hh}z.pgrb2af{fcst_hr:03d}', 'gep{mem:02d}.t{hh}z.pgrb2bf{fcst_hr:03d}']]

Traceback (most recent call last):
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/retrieve_data.py", line 1047, in <module>
    main(sys.argv[1:])
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/retrieve_data.py", line 880, in main
    write_summary_file(cla, data_store, file_templates)
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/retrieve_data.py", line 712, in write_summary_file
    [fill_template(t, cla.cycle_date, fcst_hr=fh, mem=cla.members[0]) for fh in cla.fcst_hrs]
  File "/mnt/lfs4/HFIP/hfv3gfs/Michael.Lueken/ufs-srweather-app/ush/retrieve_data.py", line 712, in <listcomp>
    [fill_template(t, cla.cycle_date, fcst_hr=fh, mem=cla.members[0]) for fh in cla.fcst_hrs]
IndexError: list index out of range
+ print_err_msg_exit 'Call to retrieve_data.py failed with a non-zero exit status.
```
I would like to successfully test at least one of these configurations before approving the changes.  Thanks!

File: ush/retrieve_data.py [hunk 7]
```diff
@@ -226,13 +226,14 @@ def fill_template(template_str, cycle_date, templates_only=False, **kwargs):
...
 
-def create_target_path(target_path):
+def create_target_path(target_path, cla, mem):
 
     """
     Append target path and create directory for ensemble members
     """
     if not os.path.exists(target_path):
         os.makedirs(target_path)
+    setattr(cla, 'output_path_{0}'.format(mem), target_path)
```
Comment (christinaholtNOAA) [inline ush/retrieve_data.py]: The modification here shouldn't be necessary given the mods below, whether the suggested change is adopted or not.

I think it night be best to leave this one with its one job.

File: ush/retrieve_data.py [hunk 8]
```diff
@@ -519,8 +523,11 @@ def hpss_requested_files(cla, file_names, store_specs, members=-1, ens_group=-1)
                 cla.cycle_date,
                 mem=mem,
             )
-
-            output_path = fill_template(cla.output_path, cla.cycle_date, mem=mem)
+            if mem != -1:
+                output_path = f"{cla.output_path}/mem{mem:03d}"
+            else:
+                output_path = cla.output_path
+            output_path = fill_template(output_path, cla.cycle_date, mem=mem)
```
Comment (christinaholtNOAA) [inline ush/retrieve_data.py]: I think this mod also goes away if the suggested change above is adopted.

File: ush/retrieve_data.py [hunk 9]
```diff
@@ -699,24 +706,48 @@ def write_summary_file(cla, data_store, file_templates):
...
+                files.extend(
+                    [fill_template(t, cla.cycle_date, fcst_hr=fh, mem=cla.members[0]) for fh in cla.fcst_hrs]
+                )
+        else:
+            files.extend(
+                [fill_template(tmpl, cla.cycle_date, fcst_hr=fh) for fh in cla.fcst_hrs]
+            )
+    
+    if isinstance(cla.members, list):
+        for m in cla.members:
```
Comment (christinaholtNOAA) [inline ush/retrieve_data.py]: Consider using `mem` here so the code is more readable.

File: ush/retrieve_data.py [hunk 10]
```diff
@@ -699,24 +706,48 @@ def write_summary_file(cla, data_store, file_templates):
...
+            for t in tmpl:
+                files.extend(
+                    [fill_template(t, cla.cycle_date, fcst_hr=fh, mem=cla.members[0]) for fh in cla.fcst_hrs]
+                )
+        else:
+            files.extend(
+                [fill_template(tmpl, cla.cycle_date, fcst_hr=fh) for fh in cla.fcst_hrs]
+            )
+    
+    if isinstance(cla.members, list):
```
Comment (christinaholtNOAA) [inline ush/retrieve_data.py]: I think that with the way the members arg is set up, this will always be `True`.

This code now contains a bunch of duplicate information and lines. I think it can be collapsed if the treatment of `summary_fp` is handled inside the logic.

File: ush/test_retrieve_data.py [hunk 11]
```diff
@@ -113,7 +113,7 @@ def test_gdas_ics_from_aws(self):
                 '--data_stores', 'aws',
                 '--external_model', 'GDAS',
                 '--fcst_hrs', '6', '9', '3',
-                '--output_path', out_path_tmpl,
+                '--output_path', tmp_dir,
```
Comment (christinaholtNOAA) [inline ush/test_retrieve_data.py]: Here, it might be good to keep `out_path_tmpl` but to join it with `tmp_dir`...this is what I was thinking...

```
out_path_templ = os.path.join(tmp_dir, f"mem{{mem:03d}}")
```

**Review by christinaholtNOAA — CHANGES_REQUESTED** [review]
@EdwardSnyder-NOAA I am sorry for my delayed review. I've left a few comments below, but overall I think this is really nice work! Thank you!

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 1 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_get_extrn_mdl_files.sh]: Yep, that was a mistake. At the time I only had access to AWS, so I had to trick the script to default to sfc only files. I removed this logic with the latest commit. Resolving.

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 2 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_get_extrn_mdl_files.sh]: Thanks for the suggestions! I'll update those changes soon. And yeah, I thought 'ary' was a bit weird but wanted to be cool and abbreviate array. haha

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 3 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_get_extrn_mdl_files.sh]: It does make sense to read the info in from the data locations file, but I couldn't come up with a way I liked to read them in. What I landed on was to use the grep command and grab the data locations that way. I'll be updating this file soon so you can see what I mean.

File: scripts/exregional_make_ics.sh (hunk 4 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_make_ics.sh:422]: I'm not sure how I set that to FV3GFS. Maybe I was looking at an older version? Regardless, I changed it to GFS.

File: scripts/exregional_make_lbcs.sh (hunk 5 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_make_lbcs.sh:302]: I only see a 'gaussian_netcdf' as the input type for GDAS in the [RRFS_dev1](https://github.com/NOAA-GSL/regional_workflow/blob/feature/RRFS_dev1/scripts/exregional_make_lbcs.sh) file. Maybe I'm not looking at the right location??

File: ush/retrieve_data.py (hunk 6 above)
Comment (EdwardSnyder-NOAA) [reply ush/retrieve_data.py]: Making this modification made updating this file MUCH easier! And initially my intuition was to pass in a file path similar to this but took the description of the output path existing too literally and abandoned the idea. Making this modification resolves the other conversations for this script.

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 3 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_get_extrn_mdl_files.sh]: @christinaholtNOAA - Where you wanting to read the file names from the summary file? Because I thought about doing it that way too but ultimately decided to read from the source (data_locations.yaml) instead.

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 12]
```diff
@@ -171,6 +182,86 @@ ${cmd}
...
+#
+function filename_mod() {
+    # Function that modifies the filenames
+    fn_mod=$(echo "$1" |
+             sed "s|{mem:02d}|$num|g" |
+             sed "s|t{hh}|t${hh}|g" |
+             sed "s|f{fcst_hr:03d}|f`printf %03d $fcst_hr`|g" |
+             sed "s|f{fcst_hr:02d}|f`printf %02d $fcst_hr`|g" )
+    echo $fn_mod
+}
```
Comment (christinaholtNOAA) [inline scripts/exregional_get_extrn_mdl_files.sh]: This use of `sed` will likely break functionality on a mac. Not sure if that matters here?

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 13]
```diff
@@ -171,6 +182,86 @@ ${cmd}
...
+    filenames=(2 4)
+    fn_list=()
+    for line in "${filenames_lines[@]}"; do
+        for name in "${filenames[@]}"; do
+            filename=$( grep -A$line 'GEFS' ${PARMdir}/data_locations.yml |
+                        tail -n1 |
+                        awk -F "'" "{ print $ $name }" )
+            fn_list+=( "$filename" )
+        done
+    done
```
Comment (christinaholtNOAA) [inline scripts/exregional_get_extrn_mdl_files.sh]: Perhaps it's easier to get the cycle-specific filenames that were downloaded from the summary file? They're already in a bash array, IIR.

This solution imposes a specified arbitrary order on the items that are expected in the data_locations.yml file that was not a requirement before.

File: scripts/exregional_make_lbcs.sh (hunk 5 above)
Comment (christinaholtNOAA) [reply scripts/exregional_make_lbcs.sh:302]: Right, but there are 2 entries present at the file at that link for `fn_atm_nemsio` and `fn_sfc_nemsio` to support nemsio input files from GDAS.

Those entries are also present in the case statement below, too.

**Review by christinaholtNOAA — CHANGES_REQUESTED** [review]
@EdwardSnyder-NOAA Thanks for addressing the changes from the last review. I left just a couple of additional comments/concerns below.

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 12 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_get_extrn_mdl_files.sh]: That function will go away with my latest changes. But I do have a sed command later in the script, which I changed to 'awk' instead.

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 13 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_get_extrn_mdl_files.sh]: I switched to using the filenames found in the summary file. Let me know what you think.

File: scripts/exregional_make_lbcs.sh (hunk 5 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_make_lbcs.sh:302]: Okay, I gotcha. In the [develop branch](https://github.com/ufs-community/ufs-srweather-app/blob/develop/scripts/exregional_make_lbcs.sh#L157), `fn_atm_nemsio` was replaced with `fn_atm` variable that accepts both nemsio and netcdf format. But to handle for a nemsio case, I'd have to modify the code block to look something like this:
```
"GDAS")
    if [ "${FV3GFS_FILE_FMT_LBCS}" = "nemsio" ]; then
        fn_atm="${EXTRN_MDL_FNS[$i]}"
    elif [ "${FV3GFS_FILE_FMT_LBCS}" = "netcdf" ]; then
        fn_atm="${EXTRN_MDL_FNS[$i]}"
```
Is this what you are looking for? My apologies for not fully understanding this request.

Comment (christinaholtNOAA) [reply scripts/exregional_make_lbcs.sh:302]: No, sorry, I should have been more explicit. In the [RRFS_dev1 version](https://github.com/NOAA-GSL/regional_workflow/blob/feature/RRFS_dev1/scripts/exregional_make_lbcs.sh), under GDAS on line 337, there are these entries:

```
"GDASENKF")
  tracers_input="[\"spfh\",\"clwmr\",\"o3mr\",\"icmr\",\"rwmr\",\"snmr\",\"grle\"]"
  tracers="[\"sphum\",\"liq_wat\",\"o3mr\",\"ice_wat\",\"rainwat\",\"snowwat\",\"graupel\"]"
  external_model="GFS"
  input_type="gaussian_netcdf"
  fn_atm_nemsio="${EXTRN_MDL_FNS[0]}"
  fn_sfc_nemsio="${EXTRN_MDL_FNS[1]}"
  ;;
```

In your addition of GDAS here, you've added a single `fn_atm` entry where the RRFS version has two lines `fn_atm_nemsio` and `fn_sfc_nemsio`. This means we can't use nemsio files for GDAS, we'd have to use netcdf. We need an option to use nemsio.

**Review by christinaholtNOAA — APPROVED** [review]
@EdwardSnyder-NOAA This looks great! Thanks for your mods!

I think I am fine with merging if you want to just open an Issue in our RRFS project board for supporting the nemsio GDAS files as I commented below. We can handle that with a much smaller PR later.

**Comment by mkavulich** [comment]
@EdwardSnyder-NOAA I am working through a code review and it looks good so far, but I wanted to bring this up early: does it make sense to add workflow end-to-end tests (or modify current ones) to include these new input data types? We could stage some static data for testing on platforms that do not have HPSS access.

Having config files for these WE2E tests would have the added benefit of being a convenient example case for anyone wanting to try out this capability.

**Comment by EdwardSnyder-NOAA** [comment]
> @EdwardSnyder-NOAA I am working through a code review and it looks good so far, but I wanted to bring this up early: does it make sense to add workflow end-to-end tests (or modify current ones) to include these new input data types? We could stage some static data for testing on platforms that do not have HPSS access.
> 
> Having config files for these WE2E tests would have the added benefit of being a convenient example case for anyone wanting to try out this capability.

@mkavulich - I like that idea a lot. We can modify one of the current WE2E tests. Would adding the GDAS and GEFS WE2E tests be included with this PR or could we create a GitHub issue for it and tackle it later?

**Comment by mkavulich** [comment]
@EdwardSnyder-NOAA I think it's okay to start a new issue and address this later. It's probably about time we started a conversation about paring down the number of existing tests again, because we are starting to proliferate again and there's a lot of consolidating we could do.

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 14]
```diff
@@ -136,6 +140,13 @@ if [ $SYMLINK_FIX_FILES = "TRUE" ]; then
   additional_flags="$additional_flags \
   --symlink"
 fi
+
```
Comment (mkavulich) [inline scripts/exregional_get_extrn_mdl_files.sh:143]: This logic doesn't appear to work correctly for ensemble cases. When I ran a test ensemble case (community_ensemble_008mems), the script fails with the following error:

```
Traceback (most recent call last):
  File "/mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/ufs-srweather-app/ush/retrieve_data.py", line 1020, in <module>
    main(sys.argv[1:])
  File "/mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/ufs-srweather-app/ush/retrieve_data.py", line 808, in main
    unavailable = get_requested_files(
  File "/mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/ufs-srweather-app/ush/retrieve_data.py", line 366, in get_requested_files
    target_path = fill_template(cla.output_path, cla.cycle_date, mem=mem)
  File "/mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/ufs-srweather-app/ush/retrieve_data.py", line 226, in fill_template
    return template_str.format(**format_values)
ValueError: Unknown format code 'd' for object of type 'str'
+ print_err_msg_exit 'Call to retrieve_data.py failed with a non-zero exit status.

The command was:

python3 -u /mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/ufs-srweather-app/ush/retrieve_data.py   --debug   --anl_or_fcst anl   --config /mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/ufs-srweather-app/parm/data_locations.yml   --cycle_date 2019070100   --data_stores disk disk   --external_model FV3GFS   --fcst_hrs 0   --ics_or_lbcs ICS   --output_path /mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/expt_dirs/community_ensemble_008mems/2019070100/FV3GFS/for_ICS/mem{mem:03d}   --summary_file extrn_mdl_var_defns.sh      --file_type nemsio   --input_file_path /mnt/lfs4/BMC/wrfruc/UFS_SRW_App/develop/input_model_data/FV3GFS/nemsio/2019070100   --symlink   --members 1 8
'
FATAL ERROR:
ERROR:
  From script:  "exregional_get_extrn_mdl_files.sh"
  Full path to script:  "/mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/ufs-srweather-app/scripts/exregional_get_extrn_mdl_files.sh"
Call to retrieve_data.py failed with a non-zero exit status.
```


Full experiment on Jet here: `/mnt/lfs4/HFIP/dtc-hurr/kavulich/workdir/PR_526/expt_dirs/community_ensemble_008mems`

**Review by mkavulich — CHANGES_REQUESTED** [review]
Hi Eddy, sorry to take so long to get you a review here. I have not yet finished my review but at least one fix needs to be made. In addition, your code is based off a quite old commit (bf07542, from mid-December), can you please update your branch to the latest `develop` branch? Thanks.

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 14 above)
Comment (EdwardSnyder-NOAA) [reply scripts/exregional_get_extrn_mdl_files.sh:143]: Hi Mike, no worries! I updated my fork branch to the latest `develop` branch. I also found a bug in the code that was causing this error (line 347 needed `members` to be defined as `cla.members`. It should run through now.

**Comment by MichaelLueken** [comment]
@EdwardSnyder-NOAA I have added the two new WE2E tests to the PR.

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 14 above)
Comment (mkavulich) [reply scripts/exregional_get_extrn_mdl_files.sh:143]: Thanks, it appears to be working now!

**Review by mkavulich — APPROVED** [review]
After the last round of changes, everything looks good, approving now.

Just a note: the GEFS and GDAS data currently only works for ensemble cases, but it's not noted explicitly in the documentation. It seems like it should be easy enough to make this data available for deterministic cases as well, but I concede that's beyond the scope of this PR. I will open an issue discussing whether we should add this capability, or just update the docs to clarify that GDAS and GEFS data is for ensemble cases only.

**Comment by venitahagerty** [comment]
Machine: hera
Compiler: intel
Job: WE
Repo location: /scratch1/BMC/zrtrr/rrfs_ci/autoci/pr/1176561214/20230210195018/ufs-srweather-app
Build was Successful
Rocoto jobs started
Long term tracking will be done on 10 experiments
If test failed, please make changes and add the following label back:
ci-hera-intel-WE
Experiment Succeeded on hera: community_ensemble_2mems_stoch
Experiment Succeeded on hera: pregen_grid_orog_sfc_climo
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional_plot
Experiment Succeeded on hera: grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
Experiment Succeeded on hera: grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
Experiment Succeeded on hera: MET_ensemble_verification
All experiments completed

**Review by MichaelLueken — APPROVED** [review]
@EdwardSnyder-NOAA These changes look good to me!  The non-Hera Jenkins tests have passed successfully and the GHA Hera tests have passed as well.  I will now approve and merge this work.
